syntax = "proto3";

package mrrpc;

service RpcAssignTask {
  // worker给master发送请求，获取任务
  rpc AssignTask(AssignTaskRequest) returns (RPCTask) {}
  // worker向master报告任务已经做完
  rpc TaskCompleted(RPCTask) returns (TaskCompletedReply) {}
}

message AssignTaskRequest {
  bytes assign_arg = 1;   // 不使用
}

message keyvalue {
  string key_value_pair = 1;
}

// 参照Task类型
message RPCTask {
  string inputs = 1;                    // 输入内容
  string outputs = 2;                   // 输出内容
  repeated keyvalue intermediates = 3;  // 中间结果，key-value对
  int32 nreducer = 4;                   // 总共的reducer数量
  int32 task_no = 5;                    // 任务编号
  int32 task_state = 6;                 // 任务所处阶段: 0-map, 1-reduce, 2-exit, 3-wait
}


message TaskCompletedReply {
  string completed_reply = 1; // 不使用
}